Code/Tools/AssetProcessor/native/unittests/FileWatcherUnitTests.cpp

```diff
@@ -5,352 +5,789 @@
+
+        m_baseFileIO = AZStd::make_unique<AZ::IO::LocalFileIO>();
+        AZ::IO::FileIOBase::SetInstance(m_baseFileIO.get());
+
+        m_tempDir = AZStd::make_unique<AZ::Test::ScopedAutoTempDirectory>();
+        // remove any symlinking.  
+        // This is necessary because on some operating systems, the temp dir
+        // may be a symlinked folder, but the file watching API generally implemented
+        // at a lower level than things like symlinks and will tend to emit
+        // real paths, which then won't match up with the expected values.
+        m_assetRootPath = QFileInfo(m_tempDir->GetDirectory()).canonicalFilePath();
+        m_fileWatcher = AZStd::make_unique<FileWatcher>();
+
+        SetupWatches();
+
+        m_fileWatcher->StartWatching();
+
+        m_addFileConnection = QObject::connect(m_fileWatcher.get(), &FileWatcher::fileAdded, [&](QString filename)
+            {
+                if (filename == m_currentFenceFilePath)
+                {
+                    m_fenceFileFound = true;
+                }
+                else
+                {
+                    m_filesAdded.push_back(filename);
+                }
+            });
+        
+        m_removeFileConnection = QObject::connect(m_fileWatcher.get(), &FileWatcher::fileRemoved, [&](QString filename)
+            {
+                m_filesRemoved.push_back(filename);
+            });
+
+        m_modifyFileConnection = QObject::connect(m_fileWatcher.get(), &FileWatcher::fileModified, [&](QString filename)
+            {
+                // its possible to get 'file modified' on a fence file, which we ignore.
+                if (filename.contains("__fence__"))
+                {
+                    return;
+                }
+                m_filesModified.insert(filename);
+            });
+
+        Flush();
+    }
 
-    m_assetRootPath = QString(m_assetDatabaseRequestsHandler->GetAssetRootDir().c_str());
-    m_fileWatcher = AZStd::make_unique<FileWatcher>();
-    m_fileWatcher->AddFolderWatch(m_assetRootPath);
-    m_fileWatcher->StartWatching();
-}
+    // default setup watches the asset root and ignores all files with the word ignored in them.
+    void FileWatcherUnitTest::SetupWatches()
+    {
+        m_fileWatcher->AddExclusion(AssetBuilderSDK::FilePatternMatcher("*ignored*", AssetBuilderSDK::AssetBuilderPattern::Wildcard));
+        m_fileWatcher->AddFolderWatch(m_assetRootPath);
+    }
 
-void FileWatcherUnitTest::TearDown()
-{
-    m_fileWatcher.reset();
+    QString FileWatcherUnitTest::GetFenceFolder()
+    {
+        return m_assetRootPath;
+    }
 
-    UnitTest::AssetProcessorUnitTestBase::TearDown();
-}
+    void FileWatcherUnitTest::CreateFenceFile()
+    {
+        AZ::IO::Path fencePath = AZ::IO::Path(GetFenceFolder().toUtf8().constData()) / "__fence__";
+        AZStd::string fenceString; 
+        ASSERT_TRUE(AZ::IO::CreateTempFileName(fencePath.String().c_str(), fenceString));
```

Comment:
You can call `.c_str()` directly on the AZ::IO::Path instance, but what is better than invoking AZ::IO::CreateTempFileName is to use an [AZ::Test::ScopedAutoTempDirectory](https://github.com/o3de/o3de/blob/development/Code/Framework/AzTest/AzTest/Utils.h#L73-L87), which creates a temporary directory that will be cleaned up at the end of the test.